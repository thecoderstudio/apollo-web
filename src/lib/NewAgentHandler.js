export default class NewAgentHandler {
  constructor() {
    this.supportedArch = ['amd64', 'arm64', 'arm'];
    this.supportedOS = ['linux', 'darwin', 'freebsd', 'openbsd'];
  };

  getDirectlyOnMachineCommand(os, arch, agentId, secret, host) {
    let command = `curl http://localhost:1970/agent/download?target_os=${os}&target_arch=${arch}`;
    return command += ` ${this.getExecuteCommand(agentId, secret, host)}`;
  };

  getExecuteCommand(agentId, secret, host) {
    return `chmod +x apollo-agent.bin && ./apollo-agent.bin --agent-id=${agentId} --secret=${secret} --host=${host}`;
  };
};