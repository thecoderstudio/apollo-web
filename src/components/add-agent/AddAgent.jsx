import React from 'react';
import styled from 'styled-components';
import axios from 'axios';
import PropTypes from 'prop-types';
import CreateAgent from './CreateAgent';
import media from '../../util/media';
import OutlinedButton from '../buttons/OutlinedButton';
import LoadingButton from '../buttons/LoadingButton';
import CopyToClipboard from '../CopyToClipboard';
import NewAgentHandler from "../../lib/NewAgentHandler";

const propTypes = {
  manualUpload: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired
};

const TwoColumnGrid = styled.div`
  display: grid;
  grid-template-columns: [column-one] 50% [column-two] 50%;
  margin: 20px 0px 20px 0px;

  ${
    media.phone`
      grid-template-columns: [column-one] 100%;
      grid-template-rows: [row-one] 1fr [row-two] 1fr;
    `
  }
`;

const ColumnOne = styled.div`
  grid-column: column-one;
  margin-right: 10px;

  ${
    media.phone`
      grid-row: row-one;
      margin-right: 0px;
      margin: 15px;
    `
  }
`;

const ColumnTwo = styled.div`
  grid-column: column-two;
  margin-left: 10px;

  ${
    media.phone`
      grid-column: column-one;
      grid-row: row-two;
      margin-left: 0px;
      margin: 15px;
    `
  }
`;

const StyledButton = styled(OutlinedButton)`
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
`;

const DownloadBinaryButton = styled(LoadingButton)`
  float: right;
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;

  ${
    media.phone`
      float: none;
      margin-left: auto;
      margin-right: auto;
    `
  }
`;

const ThreeRowDisplay = styled.div`
  display: grid;
  grid-template-rows: [description] 1fr [commands] 1fr [button] 1fr
`;

const StyledCopyToClipboard = styled(CopyToClipboard)`
  grid-row: commands;
  margin-bottom: 20px;
`;

const CloseButton = styled(StyledButton)`
  grid-row: button;
  margin-top: 20px;
`;

const Description = styled.p`
  grid-row: description;
  margin: 20px 0px 20px 0px;
`;

class AddAgent extends React.PureComponent {
	constructor(props) {
    super(props);
    this.newAgentHandler = new NewAgentHandler();
		this.state = {
      agentCreated: false,
      selectedArchitecture: null,
      selectedOperatingSystem: null
		};
	}

	downloadBinary = () => {
    this.setState({ loading: true });
    console.log(this.state);
    axios.get(
      `${process.env.APOLLO_HTTP_URL}agent/download`,
      {
        withCredentials: true,
        params: {
          target_os: this.state.selectedOperatingSystem,
          target_arch: this.state.selectedArchitecture
        },
        responseType: 'arraybuffer'
      },
    )
      .then(response => {
        this.newAgentHandler.downloadFile(response.data);
      })
      .catch(error => {
        console.log(error);
        console.log(error.response)
      })
      .finally(_ => {
        this.setState({loading: false});
      });
  };

  selectArchitecture = (value) => {
    this.setState({ selectedArchitecture: value })
  };

  selectOperatingSystem = (value) => {
    this.setState({ selectedOperatingSystem: value });
  };

  getCopyToClipboardComponents = (command) => {
		return(
			 <ThreeRowDisplay>
				 <Description>
					 Copy and run the command on the target machine to download run the client.
				 </Description>
				 <StyledCopyToClipboard id='copytoclip' text={command} />
				 <CloseButton id='closeButton' onClick={this.props.onClose}>Close</CloseButton>
			 </ThreeRowDisplay>
		 );
	 }

	createAgentSuccessCallback = (response, architecure, operatingSystem) => {
		this.setState({
      agentCreated: true,
			agentId: response.data['id'],
      secret: response.data['oauth_client']['secret'],
      selectedOperatingSystem: operatingSystem,
      selectedArchitecture: architecure
    });
	};

	getCommand = () => {
		let command;
		if (this.props.manualUpload) {
			command = this.newAgentHandler.getExecuteCommand(this.state.agentId, this.state.secret, "localhost:1970")
		} else {
			command = this.newAgentHandler.getDirectlyOnMachineCommand(
        this.props.selectedOperatingSystem, this.props.selectedArchitecture, this.state.agentId, this.state.secret,
        "localhost:1970"
      )
    }
    return command;
	};

  render() {
    return (
			<div>
				{this.state.agentCreated ?
					<div>
						{this.props.manualUpload &&
							<TwoColumnGrid>
								<ColumnOne>
									Download the binary and upload it to the target machine.
								</ColumnOne>
								<ColumnTwo>
									<DownloadBinaryButton id='downloadBinaryButton' loading={this.state.loading} onClick={this.downloadBinary}>
										Download binary
									</DownloadBinaryButton>
								</ColumnTwo>
							</TwoColumnGrid>
						}

						<ThreeRowDisplay>
							<Description>
								Copy and run the command on the target machine to download run the client.
							</Description>
							<StyledCopyToClipboard id='copytoclip' text={this.getCommand()} />
							<CloseButton id='closeButton' onClick={this.props.onClose}>Close</CloseButton>
						</ThreeRowDisplay>

					</div>
          :
          <CreateAgent createAgentSuccessCallback={this.createAgentSuccessCallback} />
				}
			</div>
    );
  }
}

AddAgent.propTypes = propTypes;

export default AddAgent;