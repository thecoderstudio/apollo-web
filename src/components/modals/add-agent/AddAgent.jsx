import React from 'react';
import styled from 'styled-components';
import axios from 'axios';
import PropTypes from 'prop-types';
import media from '../../../util/media';
import OutlinedButton from '../../buttons/OutlinedButton';
import LoadingButton from '../../buttons/LoadingButton';
import CopyToClipboard from '../../CopyToClipboard';
import NewAgentHandler from "../../../lib/NewAgentHandler";
import { handleHTTPResponse } from '../../../actions/error';

const propTypes = {
  manualUpload: PropTypes.bool.isRequired,
  selectedArchitecture: PropTypes.string.isRequired,
  selectedOperatingSystem: PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired
};

const TwoColumnGrid = styled.div`
  display: grid;
  grid-template-columns: [column-one] 50% [column-two] 50%;
  margin: 20px 0px 20px 0px;

  ${
    media.phone`
      grid-template-columns: [column-one] 100%;
      grid-template-rows: [row-one] 1fr [row-two] 1fr;
    `
  }
`;

const ColumnOne = styled.div`
  grid-column: column-one;
  margin-right: 10px;

  ${
    media.phone`
      grid-row: row-one;
      margin-right: 0px;
      margin: 15px;
    `
  }
`;

const ColumnTwo = styled.div`
  grid-column: column-two;
  margin-left: 10px;

  ${
    media.phone`
      grid-column: column-one;
      grid-row: row-two;
      margin-left: 0px;
      margin: 15px;
    `
  }
`;

const StyledButton = styled(OutlinedButton)`
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
`;

const DownloadBinaryButton = styled(LoadingButton)`
  float: right;
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;

  ${
    media.phone`
      float: none;
      margin-left: auto;
      margin-right: auto;
    `
  }
`;

const ThreeRowDisplay = styled.div`
  display: grid;
  grid-template-rows: [description] 1fr [commands] 1fr [button] 1fr
`;

const StyledCopyToClipboard = styled(CopyToClipboard)`
  grid-row: commands;
  margin-bottom: 20px;
`;

const CloseButton = styled(StyledButton)`
  grid-row: button;
  margin-top: 20px;
`;

const Description = styled.p`
  grid-row: description;
  margin: 20px 0px 20px 0px;
`;

class AddAgent extends React.PureComponent {
	constructor(props) {
    super(props);
    this.newAgentHandler = new NewAgentHandler();
    this.state = {
      loading: false
    }
	}

	downloadBinary = () => {
    this.setState({ loading: true });
    axios.get(
      `${process.env.APOLLO_HTTP_URL}agent/download`,
      {
        withCredentials: true,
        params: {
          target_os: this.props.selectedOperatingSystem,
          target_arch: this.props.selectedArchitecture
        },
        responseType: 'arraybuffer'
      },
    )
      .then(response => {
        console.log("**********")
        this.newAgentHandler.downloadFile(response.data);
      })
      .catch(error => {
        handleHTTPResponse(error.response, true, false);
      })
      .finally(_ => {
        this.setState({loading: false});
      });
  };

	getCommand = () => {
		let command;
		if (this.props.manualUpload) {
			command = this.newAgentHandler.getExecuteCommand(this.state.agentId, this.state.secret, "localhost:1970")
		} else {
			command = this.newAgentHandler.getDirectlyOnMachineCommand(
        this.props.selectedOperatingSystem, this.props.selectedArchitecture, this.state.agentId, this.state.secret,
        "localhost:1970"
      )
    }
    return command;
	};

  render() {
    return (
			<div>
        {this.props.manualUpload &&
          <TwoColumnGrid>
            <ColumnOne>
              Download the binary and upload it to the target machine.
            </ColumnOne>
            <ColumnTwo>
              <DownloadBinaryButton id='downloadBinaryButton' loading={this.state.loading} onClick={this.downloadBinary}>
                Download binary
              </DownloadBinaryButton>
            </ColumnTwo>
          </TwoColumnGrid>
        }

        <ThreeRowDisplay>
          <Description>
            Copy and run the command on the target machine to download run the client.
          </Description>
          <StyledCopyToClipboard id='copytoclip' text={this.getCommand()} />
          <CloseButton id='closeButton' onClick={this.props.onClose}>Close</CloseButton>
        </ThreeRowDisplay>
			</div>
    );
  }
}

AddAgent.propTypes = propTypes;

export default AddAgent;