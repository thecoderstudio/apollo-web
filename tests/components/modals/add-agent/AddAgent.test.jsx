import React from 'react';
import configureStore from 'redux-mock-store';
import waitForExpect from 'wait-for-expect';
import axios from 'axios';
import { StatusCodes } from 'http-status-codes';
import renderer from 'react-test-renderer';
import AddAgent from '../../../../src/components/modals/add-agent/AddAgent';
import NewAgentHandler from '../../../../src/lib/NewAgentHandler';

jest.mock('axios');

function getComponent(spy, manualUpload=false) {
  return renderer.create(
    <AddAgent
      onClose={spy}
      manualUpload={manualUpload}
      selectedArchitecture='test'
      selectedOperatingSystem='test'
    />
  );
}

describe('Add Agent', () => {
  let spy;
  let callbackSpy;

  beforeEach(() => {
    spy = jest.fn();
    callbackSpy = jest.fn();
  });

  it('renders correctly', () => {
    let tree = getComponent(spy).toJSON();
    expect(tree).toMatchSnapshot();

    tree = getComponent(spy, true).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('closes modal correctly', async () => {
    const component = getComponent(spy);
    const instance = component.root;
    instance.findByProps({ id: 'closeButton' }).props.onClick();
    expect(component.toJSON()).toMatchSnapshot();

    await waitForExpect(() => {
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  it('calls download file correctly after successful get', async () => {
    const component = getComponent(spy, true);
    const downloadFileSpy = jest.spyOn(NewAgentHandler.prototype, 'downloadFile').mockImplementation(() => { return; });

    axios.get.mockResolvedValue({
      status: 200,
      data: ''
    });
    component.root.findByProps({ id: 'downloadBinaryButton' }).props.onClick();

    await waitForExpect(() => {
      expect(downloadFileSpy).toHaveBeenCalled();
    });
  });

   it('Does not call download file on error', async () => {
      const component = getComponent(spy, true);
      const downloadFileSpy = jest.spyOn(NewAgentHandler.prototype, 'downloadFile').mockImplementation(() => { return; });

      axios.get.mockRejectedValue({
        status: 400,
        response: {
          data: {}
        }
      });

      component.root.findByProps({ id: 'downloadBinaryButton' }).props.onClick();

      await waitForExpect(() => {
        expect(downloadFileSpy).not.toHaveBeenCalled();
      });
    });
});