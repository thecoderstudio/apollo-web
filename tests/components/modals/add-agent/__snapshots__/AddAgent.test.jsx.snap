// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Agent closes modal correctly 1`] = `
.c5 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
}

.c5:hover {
  opacity: 0.90;
}

.c5:focus {
  outline: none;
}

.c5:active {
  opacity: 0.80;
}

.c5:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c8 {
  background-color: transparent;
  border: 1px solid;
}

.c2 {
  grid-template-columns: [textField] 1fr [copy] 50px;
  display: grid;
  height: 50px;
}

.c4 {
  grid-column: textField;
  border: none;
  border-radius: 5px;
  color: white;
  cursor: text;
  padding: 15px;
  overflow: auto;
  white-space: nowrap;
  overflow-y: hidden;
  overflow-x: scroll;
}

.c4::-webkit-scrollbar {
  background: transparent;
}

.c6 {
  grid-column: copy;
  height: 100%;
}

.c7 {
  text-align: center;
  display: block;
}

.c9 {
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c0 {
  display: grid;
  grid-template-rows: [description] 1fr [commands] 1fr [button] 1fr;
}

.c3 {
  grid-row: commands;
  margin-bottom: 20px;
}

.c10 {
  grid-row: button;
  margin-top: 20px;
}

.c1 {
  grid-row: description;
  margin: 20px 0px 20px 0px;
}

<div>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Copy and run the command on the target machine to download run the client.
    </p>
    <div
      className="c2 c3"
    >
      <div
        className="c4"
      >
        curl http://localhost:1234/agent/download?target_os=test&target_arch=test &gt; apollo-agent.bin && chmod +x apollo-agent.bin && ./apollo-agent.bin --agent-id=id --secret=secret --host=http://localhost:1234
      </div>
      <button
        className="c5 c6"
        id="copyToClipboardButton"
        onClick={[Function]}
      >
        <i
          className="c7 fas fa-copy"
        />
      </button>
    </div>
    <button
      className="c5 c8 c9 c10"
      id="closeButton"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    >
      Close
    </button>
  </div>
</div>
`;

exports[`Add Agent renders correctly 1`] = `
.c5 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
}

.c5:hover {
  opacity: 0.90;
}

.c5:focus {
  outline: none;
}

.c5:active {
  opacity: 0.80;
}

.c5:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c8 {
  background-color: transparent;
  border: 1px solid;
}

.c2 {
  grid-template-columns: [textField] 1fr [copy] 50px;
  display: grid;
  height: 50px;
}

.c4 {
  grid-column: textField;
  border: none;
  border-radius: 5px;
  color: white;
  cursor: text;
  padding: 15px;
  overflow: auto;
  white-space: nowrap;
  overflow-y: hidden;
  overflow-x: scroll;
}

.c4::-webkit-scrollbar {
  background: transparent;
}

.c6 {
  grid-column: copy;
  height: 100%;
}

.c7 {
  text-align: center;
  display: block;
}

.c9 {
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c0 {
  display: grid;
  grid-template-rows: [description] 1fr [commands] 1fr [button] 1fr;
}

.c3 {
  grid-row: commands;
  margin-bottom: 20px;
}

.c10 {
  grid-row: button;
  margin-top: 20px;
}

.c1 {
  grid-row: description;
  margin: 20px 0px 20px 0px;
}

<div>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Copy and run the command on the target machine to download run the client.
    </p>
    <div
      className="c2 c3"
    >
      <div
        className="c4"
      >
        curl http://localhost:1234/agent/download?target_os=test&target_arch=test &gt; apollo-agent.bin && chmod +x apollo-agent.bin && ./apollo-agent.bin --agent-id=id --secret=secret --host=http://localhost:1234
      </div>
      <button
        className="c5 c6"
        id="copyToClipboardButton"
        onClick={[Function]}
      >
        <i
          className="c7 fas fa-copy"
        />
      </button>
    </div>
    <button
      className="c5 c8 c9 c10"
      id="closeButton"
      onClick={[MockFunction]}
    >
      Close
    </button>
  </div>
</div>
`;

exports[`Add Agent renders correctly 2`] = `
.c3 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
}

.c3:hover {
  opacity: 0.90;
}

.c3:focus {
  outline: none;
}

.c3:active {
  opacity: 0.80;
}

.c3:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c13 {
  background-color: transparent;
  border: 1px solid;
}

.c4 {
  width: 100%;
  height: 50px;
  margin: auto;
  display: block;
}

.c8 {
  grid-template-columns: [textField] 1fr [copy] 50px;
  display: grid;
  height: 50px;
}

.c10 {
  grid-column: textField;
  border: none;
  border-radius: 5px;
  color: white;
  cursor: text;
  padding: 15px;
  overflow: auto;
  white-space: nowrap;
  overflow-y: hidden;
  overflow-x: scroll;
}

.c10::-webkit-scrollbar {
  background: transparent;
}

.c11 {
  grid-column: copy;
  height: 100%;
}

.c12 {
  text-align: center;
  display: block;
}

.c0 {
  display: grid;
  grid-template-columns: [column-one] 50% [column-two] 50%;
  margin: 20px 0px 20px 0px;
}

.c1 {
  grid-column: column-one;
  margin-right: 10px;
}

.c2 {
  grid-column: column-two;
  margin-left: 10px;
}

.c14 {
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c5 {
  float: right;
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c6 {
  display: grid;
  grid-template-rows: [description] 1fr [commands] 1fr [button] 1fr;
}

.c9 {
  grid-row: commands;
  margin-bottom: 20px;
}

.c15 {
  grid-row: button;
  margin-top: 20px;
}

.c7 {
  grid-row: description;
  margin: 20px 0px 20px 0px;
}

@media (max-width:47.9375em) {
  .c0 {
    grid-template-columns: [column-one] 100%;
    grid-template-rows: [row-one] 1fr [row-two] 1fr;
  }
}

@media (max-width:47.9375em) {
  .c1 {
    grid-row: row-one;
    margin-right: 0px;
    margin: 15px;
  }
}

@media (max-width:47.9375em) {
  .c2 {
    grid-column: column-one;
    grid-row: row-two;
    margin-left: 0px;
    margin: 15px;
  }
}

@media (max-width:47.9375em) {
  .c5 {
    float: none;
    margin-left: auto;
    margin-right: auto;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      Download the binary and upload it to the target machine.
    </div>
    <div
      className="c2"
    >
      <button
        className="c3 c4 c5"
        onClick={[Function]}
      >
        Download binary
      </button>
    </div>
  </div>
  <div
    className="c6"
  >
    <p
      className="c7"
    >
      Copy and run the command on the target machine to download run the client.
    </p>
    <div
      className="c8 c9"
    >
      <div
        className="c10"
      >
        chmod +x apollo-agent.bin && ./apollo-agent.bin --agent-id=id --secret=secret --host=http://localhost:1234
      </div>
      <button
        className="c3 c11"
        id="copyToClipboardButton"
        onClick={[Function]}
      >
        <i
          className="c12 fas fa-copy"
        />
      </button>
    </div>
    <button
      className="c3 c13 c14 c15"
      id="closeButton"
      onClick={[MockFunction]}
    >
      Close
    </button>
  </div>
</div>
`;
